src/module_4/push_exploration_non_linear.md

```diff
@@ -0,0 +1,2005 @@
+    importances = model.feature_importances_
+    indices = np.argsort(importances)[::-1]
+
+    plt.figure(figsize=(10, 6))
+    plt.title(title)
+    plt.bar(range(len(feature_names)), importances[indices], align="center")
+    plt.xticks(range(len(feature_names)), np.array(feature_names)[indices], rotation=90)
+    plt.xlim([-1, len(feature_names)])
+    plt.show()
+
+```
+
+
+```python
+X_train, y_train, X_val, y_val, X_test, y_test = temporal_data_split(data_processed, cols=cols)
+feature_names = X_train.columns
+```
+
+    INFO - Date from: 2020-10-05
+    INFO - Date to: 2021-03-03
+    INFO - Train set ratio: 73.44%
+    INFO - Validation set ratio: 13.32%
+    INFO - Test set ratio: 13.24%
+
+
+
+```python
+# Train a RandomForestClassifier with 200 estimators
+rf = RandomForestClassifier(n_estimators=200)
+rf.fit(X_train, y_train)
+
+# Assuming you have already defined the feature_names variable
+# If not, you can define it like this:
+# feature_names = X_train.columns
+
+# Plot feature importance for the trained RandomForest model
+plot_feature_importance(rf, feature_names, title="Random Forest (n_trees=200) Feature Importance")
+
+# Display the feature importance plot
+plt.show()
+
+```
+
+
+    
+![png](push_exploration_non_linear_files/push_exploration_non_linear_22_0.png)
+    
+
+
+
+If performance could be a significant issue in this case, we can consider discarding from 'count_pets' to the right at a later time. However, we would continue in this case with all the cols
+
+## Gradient Boosting Trees
+
+
+```python
+from sklearn.ensemble import GradientBoostingClassifier
+
+# Let's be conservative for reducing total time
+n_trees_values = [5, 100, 250]
+max_depth_values = [1, 3, 5]
+
+trained_gbt_models = []
+y_preds_gbt = []
+gbt_model_names = []
+
+for n_trees in n_trees_values:
+    for depth in max_depth_values:
+        gbt = GradientBoostingClassifier(
+            n_estimators=n_trees,
+            max_depth=depth
+            )
+        gbt.fit(X_train, y_train)
+
+        trained_gbt_models.append(gbt)
+        y_preds_gbt.append(gbt.predict_proba(X_val)[:, 1])
+        gbt_model_names.append(f"GBT (n_trees={n_trees}, depth={depth})")
+```
+
+
+```python
+lr_model_name = ["LogisticRegression"]
+rf_model_names = [f"RandomForest (n_trees={n})" for n in n_trees_values]
+
+all_model_names = lr_model_name  + gbt_model_names
+all_y_preds = [y_pred_lr]  + y_preds_gbt
+
+# Now use the generate_evaluation_curves_multi function to plot:
+generate_evaluation_curves_multi(
+    model_names=all_model_names,
+    y_preds=all_y_preds,
+    y_test=y_val
+)
+```
+
+
+    
+![png](push_exploration_non_linear_files/push_exploration_non_linear_26_0.png)
+    
+
+
+GBT with n_trees = 100 and depth = 3 seems to be our best model (brown line, legend is wrong for depth values but do not want to recompute bc of runtime) yet very similar to LR. As additional info, all models took around 30 min to run (again, if we need that info into account).
```

Comment:
all models took around 30 min to run
Again, I'm pretty sure taht can be improved